src/veranda/raster/native/netcdf.py

```diff
@@ -0,0 +1,944 @@
+                                         engine=self._engine,
+                                         use_cftime=False,
+                                         decode_cf=True,
+                                         decode_coords="all")
+
+            self.geotrans = tuple(self.src.rio.transform())
+            self.sref_wkt = self.src.rio.crs
+            self.metadata = self.src.attrs
+            self._reset()
+        elif self.mode == 'w':
+            self.src = ds
+            if self.sref_wkt is not None:
+                self.src.rio.write_crs(self.sref_wkt, inplace=True)
+            self.src.rio.write_transform(Affine(*self.geotrans), inplace=True)
+            self.src.attrs.update(self.metadata)
+            self._reset()
+            n_rows, n_cols = self.raster_shape
+            ds[self._space_dims[0]] = self.geotrans[3] + \
+                      (0.5 + np.arange(n_rows)) * self.geotrans[4] + \
+                      (0.5 + np.arange(n_rows)) * self.geotrans[5]
+            ds[self._space_dims[1]] = self.geotrans[0] + \
+                   (0.5 + np.arange(n_cols)) * self.geotrans[1] + \
+                   (0.5 + np.arange(n_cols)) * self.geotrans[2]
+        else:
+            err_msg = f"Mode '{self.mode}' not known."
+            raise ValueError(err_msg)
+
+    def _reset(self):
+        """ Resets internal class variables with properties from an existing NetCDF dataset. """
+        stack_dims = list(set(self.src.dims.keys()) - set(self._space_dims))
+        if self.mode == 'r':
+            self._stack_dims = dict()
+            for stack_dim in stack_dims:
+                self._stack_dims[stack_dim] = len(self.src[stack_dim])
+
+        dims = stack_dims + self._space_dims
+        if len(self.data_variables) == 0:
+            self.data_variables = [self.src[dvar].name for dvar in self.src.data_vars
+                                   if list(self.src[dvar].dims) == dims]
+        all_variables = self.data_variables + dims
+        for variable in all_variables:
+            self._compressions[variable] = self._compressions.get(variable, None)
+            ref_chunksizes = self.src[variable].encoding.get('chunksizes', None)
+            self._chunksizes[variable] = self._chunksizes.get(variable, ref_chunksizes)
+            if variable in self.data_variables:
+                ref_scale_factor = self.scale_factors.get(variable, 1)
+                self.scale_factors[variable] = self.src[variable].attrs.get('scale_factor', ref_scale_factor)
+                ref_offset = self.offsets.get(variable, 0)
+                self.offsets[variable] = self.src[variable].attrs.get('add_offset', ref_offset)
+                ref_nodataval = self.nodatavals.get(variable, 127)
+                self.nodatavals[variable] = self.src[variable].attrs.get('_FillValue', ref_nodataval)
+                ref_dtype = self.dtypes.get(variable, self.src[variable].dtype.name)
+                self.dtypes[variable] = ref_dtype
```

Comment:
extract

Opportunity to reduce code duplication - I've seen similar code before

Reply:
Done.